### PR TITLE
codegen-kotlin: Fix generation of copy() methods

### LINKED ISSUE #569: Koltin generator fails with "'copy' overrides nothing." for subclasses of abstract class
Try generating Kotlin classes for this template:
```
abstract class Day {
  day: Int
  month: Int
  year: Int
}

class Weekend extends Day {}
class Workday extends Day {}

days: Listing<Day>
```

The classes `Weekend` and `Workday` contain copy methods:
```
    override fun copy(
      day: Long,
      month: Long,
      year: Long
    ): Workday = Workday(day, month, year)
```
but the class `Day` which they extend:
```
class Workday(
    day: Long,
    month: Long,
    year: Long
  ) : Day(day, month, year) {
```
does not define `copy` method, which leads to a compilation error:
```
e: file:///home/whoever/whatever/app/build/generated/pkl/template/kotlin/Calendar.kt:52:5 'copy' overrides nothing.
```

It looks like the logic to [skip generation](https://github.com/apple/pkl/blob/5cc2ea2d003a1e2c6e856363684e2aa085ed38a9/pkl-codegen-kotlin/src/main/kotlin/org/pkl/codegen/kotlin/KotlinCodeGenerator.kt#L282) of `copy` methods for abstract classes is flawed.

When making the hierarchy deeper:
```
open class TimeInterval {}

abstract  class Day extends TimeInterval {
...
```
it even fails to build with "Cannot create an instance of an abstract class." (apart from `'copy' overrides nothing.`). This case should also be considered when fixing the implementation of `copy`.

Marking `Day` as `open` instead of `abstract` can be used to work around this issue.

### PR DESCRIPTION
**Pull request by translatenix**
- don't generate copy methods for abstract classes
- precisely determine if copy method needs override modifier (tricky)

Fixes #569

### CODE REVIEW

File: pkl-codegen-kotlin/src/test/kotlin/org/pkl/codegen/kotlin/KotlinCodeGeneratorTest.kt [hunk 1]
```diff
@@ -661,6 +661,184 @@ class KotlinCodeGeneratorTest {
     assertCompilesSuccessfully(kotlinCode)
   }
 
+  // https://github.com/apple/pkl/issues/569
+  @Test
+  fun abstractClasses() {
```
Comment (bioball) [inline pkl-codegen-kotlin/src/test/kotlin/org/pkl/codegen/kotlin/KotlinCodeGeneratorTest.kt:666]:
```suggestion
  fun `abstract classes`() {
```

File: pkl-codegen-kotlin/src/test/kotlin/org/pkl/codegen/kotlin/KotlinCodeGeneratorTest.kt [hunk 2]
```diff
@@ -661,6 +661,184 @@ class KotlinCodeGeneratorTest {
...
+        .trimMargin(),
+      kotlinCode
+    )
+
+    assertCompilesSuccessfully(kotlinCode)
+  }
+
+  // https://github.com/apple/pkl/issues/569
+  @Test
+  fun abstractClassThatExtendsOpenClass() {
```
Comment (bioball) [inline pkl-codegen-kotlin/src/test/kotlin/org/pkl/codegen/kotlin/KotlinCodeGeneratorTest.kt:727]:
```suggestion
  fun `abstract class that extends open class`() {
```

File: pkl-codegen-kotlin/src/test/kotlin/org/pkl/codegen/kotlin/KotlinCodeGeneratorTest.kt [hunk 3]
```diff
@@ -661,6 +661,184 @@ class KotlinCodeGeneratorTest {
...
+        .trimMargin(),
+      kotlinCode
+    )
+
+    assertCompilesSuccessfully(kotlinCode)
+  }
+
+  // https://github.com/apple/pkl/issues/569
+  @Test
+  fun abstractClassThatExtendsOpenClassWithoutAddingProperties() {
```
Comment (bioball) [inline pkl-codegen-kotlin/src/test/kotlin/org/pkl/codegen/kotlin/KotlinCodeGeneratorTest.kt:791]:
```suggestion
  fun `abstract class that extends open class without adding properties`() {
```

File: pkl-codegen-kotlin/src/main/kotlin/org/pkl/codegen/kotlin/KotlinCodeGenerator.kt [hunk 4]
```diff
@@ -274,9 +274,27 @@ class KotlinCodeGenerator(
...
+    fun nearestNonAbstractAncestorDeclaresCopyMethodWithSameArity(): Boolean {
+      if (properties.isNotEmpty()) {
+        // assertion: pClass declares properties that are generated
+        return false
+      }
+      for (currClass in generateSequence(pClass.superclass) { it.superclass }) {
+        if (!currClass.isAbstract) return true
+        if (currClass.properties.values.any { !it.isHidden }) {
+          // assertion: currClass declares properties that are generated
+          return false
```
Comment (bioball) [inline pkl-codegen-kotlin/src/main/kotlin/org/pkl/codegen/kotlin/KotlinCodeGenerator.kt:286]: Does this ever get checked? I put a breakpoint here and ran the entire test suite, and never hit this line. Plus, with your new change, we now don't generate copy methods for abstract classes, right?

File: pkl-codegen-kotlin/src/main/kotlin/org/pkl/codegen/kotlin/KotlinCodeGenerator.kt [hunk 5]
```diff
@@ -285,7 +303,8 @@ class KotlinCodeGenerator(
 
         // avoid generating multiple methods with same no. of parameters
         if (currParameters.size < prevParameterCount) {
-          val isOverride = currClass !== pClass || superclass != null && properties.isEmpty()
+          val isOverride =
+            currClass !== pClass || nearestNonAbstractAncestorDeclaresCopyMethodWithSameArity()
```
Comment (bioball) [inline pkl-codegen-kotlin/src/main/kotlin/org/pkl/codegen/kotlin/KotlinCodeGenerator.kt:307]: [nit] Either make `nearestNonAbstractAncestorDeclaresCopyMethodWithSameArity` a lazy value, or cache this so we don't call it on every iteration

**Review by bioball — CHANGES_REQUESTED** [review]
Just a few nits and comments, but otherwise looks good!

File: pkl-codegen-kotlin/src/main/kotlin/org/pkl/codegen/kotlin/KotlinCodeGenerator.kt (hunk 4 above)
Comment (translatenix) [reply pkl-codegen-kotlin/src/main/kotlin/org/pkl/codegen/kotlin/KotlinCodeGenerator.kt:286]: Good catch. This implementation is neither obvious nor correct, so I rewrote it.

File: pkl-codegen-kotlin/src/main/kotlin/org/pkl/codegen/kotlin/KotlinCodeGenerator.kt (hunk 5 above)
Comment (translatenix) [reply pkl-codegen-kotlin/src/main/kotlin/org/pkl/codegen/kotlin/KotlinCodeGenerator.kt:307]: It's only called once due to `currClass !== pClass ||`.

**Comment by translatenix** [comment]
@bioball ready from my side

File: pkl-codegen-kotlin/src/main/kotlin/org/pkl/codegen/kotlin/KotlinCodeGenerator.kt (hunk 5 above)
Comment (bioball) [reply pkl-codegen-kotlin/src/main/kotlin/org/pkl/codegen/kotlin/KotlinCodeGenerator.kt]: Ah, yeah, that was a little confusing, but you're right

**Review by bioball — APPROVED** [review]
LGTM, thank you for the fix!

**Pull request merged**